File: src/semantic_references/catalog.py
"""Parsing and structural validation of ``references/sources.toml``.

The catalog is human-authored research intent (design spec 0004, "Catalog
boundary"). Parsing here never touches Git or the network: everything
checkable from the TOML text alone is checked here, and only here.
"""

from __future__ import annotations

import hashlib
import json
import re
import tomllib
from dataclasses import dataclass, field
from pathlib import Path, PurePosixPath
from typing import cast

from semantic_references.errors import CatalogError

_ID_PATTERN = re.compile(r"^[a-z0-9][a-z0-9-]*(\.[a-z0-9][a-z0-9-]*)*$")
_CONTROL_CHAR_MAX = 0x1F
_DEL_CHAR = 0x7F


def is_git_safe_value(value: str) -> bool:
    """Reject values that could be misread as a CLI option or carry control bytes.

    Applied to every catalog/lock field that is ever passed as a bare
    positional argument to ``git`` (origin URLs, refs, aliases): a leading
    ``-`` risks being parsed as an option, and control characters (including
    ANSI escapes) have no legitimate place in a Git ref or URL.
    """
    if not value:
        return False
    if value.startswith("-"):
        return False
    return not any(ord(ch) <= _CONTROL_CHAR_MAX or ord(ch) == _DEL_CHAR for ch in value)


@dataclass(frozen=True, slots=True)
class CatalogSource:
    """One ``[[source]]`` record from ``references/sources.toml``."""

    id: str
    kind: str
    origin: str
    local_hint: str | None
    origin_aliases: tuple[str, ...]
    track: str | None
    license_paths: tuple[str, ...]
    classes: tuple[str, ...]
    questions: tuple[str, ...]
    raw: dict[str, object] = field(repr=False)

    @property
    def lockable(self) -> bool:
        """A source is lockable once both custody fields are declared."""
        return self.track is not None and len(self.license_paths) > 0

    def canonical_digest(self) -> str:
        """SHA-256 of the complete canonical catalog record.

        Canonical form is the raw TOML record re-serialized as JSON with
        recursively sorted object keys and compact separators, so the digest
        is stable across whitespace/comment/key-order changes but sensitive
        to any field or value change (design spec 0004, ``reference-lock-v1``).
        """
        canonical = json.dumps(self.raw, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
        return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


@dataclass(frozen=True, slots=True)
class Catalog:
    """The full, validated set of catalog sources, keyed by ID."""

    sources: dict[str, CatalogSource]

    def get(self, source_id: str) -> CatalogSource | None:
        return self.sources.get(source_id)


def _require_str(record: dict[str, object], key: str, source_id: str) -> str:
    value = record.get(key)
    if not isinstance(value, str) or not value:
        raise CatalogError(f"source {source_id!r}: field {key!r} must be a non-empty string")
    return value


def _optional_str(record: dict[str, object], key: str, source_id: str) -> str | None:
    if key not in record:
        return None
    value = record[key]
    if not isinstance(value, str) or not value:
        raise CatalogError(f"source {source_id!r}: field {key!r} must be a non-empty string")
    return value


def _optional_str_list(record: dict[str, object], key: str, source_id: str) -> tuple[str, ...]:
    if key not in record:
        return ()
    value = record[key]
    if not isinstance(value, list):
        raise CatalogError(f"source {source_id!r}: field {key!r} must be a list of strings")
    raw_items = cast("list[object]", value)
    if not all(isinstance(item, str) for item in raw_items):
        raise CatalogError(f"source {source_id!r}: field {key!r} must be a list of strings")
    items = [item for item in raw_items if isinstance(item, str)]
    if not items:
        raise CatalogError(f"source {source_id!r}: field {key!r} must not be an empty list")
    return tuple(items)


def validate_source_id(source_id: str) -> None:
    if not _ID_PATTERN.match(source_id):
        raise CatalogError(
            f"source id {source_id!r} is not a path-safe dotted identifier "
            "(expected lowercase alphanumeric/hyphen segments joined by single dots)"
        )


def validate_license_path(source_id: str, raw_path: str) -> str:
    """Normalize and validate one declared license artifact path.

    Must be relative, use forward slashes, contain no ``.``/``..`` segments,
    and carry no leading/trailing slash.
    """
    if not raw_path or raw_path != raw_path.strip():
        raise CatalogError(f"source {source_id!r}: license path {raw_path!r} is not normalized")
    if "\\" in raw_path:
        raise CatalogError(f"source {source_id!r}: license path {raw_path!r} must use '/'")
    if raw_path.startswith("/") or raw_path.endswith("/"):
        raise CatalogError(
            f"source {source_id!r}: license path {raw_path!r} must be relative and not "
            "trail a slash"
        )
    pure = PurePosixPath(raw_path)
    parts = pure.parts
    if not parts or any(part in ("", ".", "..") for part in parts):
        raise CatalogError(
            f"source {source_id!r}: license path {raw_path!r} must be a normalized relative path"
        )
    if str(pure) != raw_path:
        raise CatalogError(f"source {source_id!r}: license path {raw_path!r} is not normalized")
    return raw_path


def _validate_source(record: dict[str, object]) -> CatalogSource:
    raw_id = record.get("id")
    if not isinstance(raw_id, str) or not raw_id:
        raise CatalogError("source record is missing a non-empty string 'id'")
    source_id = raw_id
    validate_source_id(source_id)

    kind = _require_str(record, "kind", source_id)
    origin = _require_str(record, "origin", source_id)
    if not is_git_safe_value(origin):
        raise CatalogError(
            f"source {source_id!r}: 'origin' is not safe (option-like or has control characters)"
        )
    local_hint = _optional_str(record, "local_hint", source_id)
    origin_aliases = _optional_str_list(record, "origin_aliases", source_id)
    for alias in origin_aliases:
        if not is_git_safe_value(alias):
            raise CatalogError(
                f"source {source_id!r}: origin_aliases entry {alias!r} is not safe "
                "(option-like or has control characters)"
            )
    track = _optional_str(record, "track", source_id)
    if track is not None and not is_git_safe_value(track):
        raise CatalogError(
            f"source {source_id!r}: 'track' is not safe (option-like or has control characters)"
        )
    license_paths_raw = _optional_str_list(record, "license_paths", source_id)
    classes = _optional_str_list(record, "classes", source_id)
    questions = _optional_str_list(record, "questions", source_id)

    license_paths = tuple(validate_license_path(source_id, path) for path in license_paths_raw)
    if len(set(license_paths)) != len(license_paths):
        raise CatalogError(f"source {source_id!r}: license_paths contains duplicates")

    if (track is None) != (len(license_paths) == 0):
        raise CatalogError(
            f"source {source_id!r}: 'track' and 'license_paths' must be declared together "
            "(a source is either fully lockable or fully unlocked)"
        )

    return CatalogSource(
        id=source_id,
        kind=kind,
        origin=origin,
        local_hint=local_hint,
        origin_aliases=origin_aliases,
        track=track,
        license_paths=license_paths,
        classes=classes,
        questions=questions,
        raw=record,
    )


def parse_catalog_text(text: str) -> Catalog:
    try:
        document: dict[str, object] = tomllib.loads(text)
    except tomllib.TOMLDecodeError as exc:
        raise CatalogError(f"catalog is not valid TOML: {exc}") from exc

    schema = document.get("schema")
    if schema != 1:
        raise CatalogError(f"catalog schema {schema!r} is not the supported value (1)")

    records_field = document.get("source", [])
    if not isinstance(records_field, list):
        raise CatalogError("catalog field 'source' must be an array of tables")
    records = cast("list[object]", records_field)

    sources: dict[str, CatalogSource] = {}
    for record in records:
        if not isinstance(record, dict):
            raise CatalogError("each [[source]] entry must be a table")
        typed_record = cast("dict[str, object]", record)
        source = _validate_source(typed_record)
        if source.id in sources:
            raise CatalogError(f"duplicate source id {source.id!r}")
        sources[source.id] = source

    return Catalog(sources=sources)


def load_catalog(path: Path) -> Catalog:
    try:
        text = path.read_text(encoding="utf-8")
    except OSError as exc:
        raise CatalogError(f"cannot read catalog at {path}: {exc}") from exc
    return parse_catalog_text(text)

File: src/references/catalog.ts
import { Crypto, Effect, FileSystem, Schema } from "effect";
import { CatalogError } from "./errors.ts";
import { TomlParser } from "./toml.ts";

const ID_PATTERN = /^[a-z0-9][a-z0-9-]*(\.[a-z0-9][a-z0-9-]*)*$/;
const CONTROL_CHAR_MAX = 0x1f;
const DEL_CHAR = 0x7f;

/**
 * Reject values that could be misread as a CLI option or carry control
 * bytes. Applied to every catalog/lock field that is ever passed as a bare
 * positional argument to `git` (origin URLs, refs, aliases): a leading `-`
 * risks being parsed as an option, and control characters (including ANSI
 * escapes) have no legitimate place in a Git ref or URL.
 */
export const isGitSafeValue = (value: string): boolean => {
  if (value.length === 0 || value.startsWith("-")) return false;
  for (const ch of value) {
    const code = ch.codePointAt(0)!;
    if (code <= CONTROL_CHAR_MAX || code === DEL_CHAR) return false;
  }
  return true;
};

const REF_FORBIDDEN = new Set([" ", "~", "^", ":", "?", "*", "[", "\\"]);

/**
 * Whether `value` is a concrete, syntactically valid `refs/...` name: the
 * validation subset of `git check-ref-format` needed at the pure
 * catalog/lock boundary. Selectors such as `HEAD` and short branch names are
 * intentionally excluded — a persisted resolution must name the concrete
 * advertised ref.
 */
export const isConcreteGitRef = (value: string): boolean => {
  if (!value.startsWith("refs/") || !isGitSafeValue(value)) return false;
  if (
    value.endsWith("/") ||
    value.endsWith(".") ||
    value.includes("//") ||
    value.includes("..") ||
    value.includes("@{")
  ) {
    return false;
  }
  for (const ch of value) if (REF_FORBIDDEN.has(ch)) return false;
  const components = value.split("/");
  return components.every(
    (component) =>
      component.length > 0 && !component.startsWith(".") && !component.endsWith(".lock"),
  );
};

export const isValidSourceId = (sourceId: string): boolean => ID_PATTERN.test(sourceId);

export const validateSourceId = (sourceId: string): Effect.Effect<string, CatalogError> =>
  isValidSourceId(sourceId)
    ? Effect.succeed(sourceId)
    : Effect.fail(
        new CatalogError({
          message:
            `source id ${JSON.stringify(sourceId)} is not a path-safe dotted identifier ` +
            "(expected lowercase alphanumeric/hyphen segments joined by single dots)",
        }),
      );

/**
 * Normalized, relative, `.`/`..`-free, forward-slash license artifact path
 * (mirrors the validation subset of Python's `PurePosixPath` normalization
 * check: reject anything whose reconstruction from its segments differs from
 * the original text).
 */
export const isValidLicensePath = (rawPath: string): boolean => {
  if (rawPath.length === 0 || rawPath !== rawPath.trim()) return false;
  if (rawPath.includes("\\")) return false;
  if (rawPath.startsWith("/") || rawPath.endsWith("/")) return false;
  const parts = rawPath.split("/");
  if (parts.some((part) => part === "" || part === "." || part === "..")) return false;
  return parts.join("/") === rawPath;
};

export const validateLicensePath = (rawPath: string): Effect.Effect<string, CatalogError> =>
  isValidLicensePath(rawPath)
    ? Effect.succeed(rawPath)
    : Effect.fail(
        new CatalogError({ message: `license path ${JSON.stringify(rawPath)} is not normalized` }),
      );

export interface CatalogSource {
  readonly id: string;
  readonly kind: string;
  readonly origin: string;
  readonly localHint: string | null;
  readonly originAliases: ReadonlyArray<string>;
  readonly track: string | null;
  readonly licensePaths: ReadonlyArray<string>;
  readonly classes: ReadonlyArray<string>;
  readonly questions: ReadonlyArray<string>;
  /** The untouched TOML record, used verbatim as the canonical-digest input. */
  readonly raw: Readonly<Record<string, unknown>>;
}

/** A source is lockable once both custody fields are declared. */
export const isLockable = (source: CatalogSource): boolean =>
  source.track !== null && source.licensePaths.length > 0;

const escapeJsonString = (value: string): string => {
  let out = '"';
  for (let index = 0; index < value.length; index += 1) {
    const ch = value[index]!;
    const code = value.charCodeAt(index);
    switch (ch) {
      case '"':
        out += '\\"';
        break;
      case "\\":
        out += "\\\\";
        break;
      case "\b":
        out += "\\b";
        break;
      case "\f":
        out += "\\f";
        break;
      case "\n":
        out += "\\n";
        break;
      case "\r":
        out += "\\r";
        break;
      case "\t":
        out += "\\t";
        break;
      default:
        out += code < 0x20 || code > 0x7e ? `\\u${code.toString(16).padStart(4, "0")}` : ch;
    }
  }
  return out + '"';
};

/**
 * `JSON.stringify(value, Object.keys(value).sort())`-shaped output but
 * matching Python's `json.dumps(value, sort_keys=True, separators=(",",
 * ":"), ensure_ascii=True)` byte-for-byte: compact separators, keys sorted
 * recursively, and every non-ASCII/control character escaped as `\uXXXX`
 * (explicitly iterating UTF-16 code units makes astral characters produce
 * the same surrogate-pair escape sequence Python emits).
 */
const canonicalizeForDigest = (value: unknown): string => {
  if (value === null) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new Error("canonical digest rejects non-finite numbers");
    }
    return String(value);
  }
  if (typeof value === "string") return escapeJsonString(value);
  if (Array.isArray(value)) return `[${value.map(canonicalizeForDigest).join(",")}]`;
  if (typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>).sort(([left], [right]) =>
      left < right ? -1 : left > right ? 1 : 0,
    );
    return `{${entries.map(([key, item]) => `${escapeJsonString(key)}:${canonicalizeForDigest(item)}`).join(",")}}`;
  }
  throw new Error(`canonical digest cannot encode a value of type ${typeof value}`);
};

const toHex = (bytes: Uint8Array): string =>
  Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");

/** SHA-256 of the complete canonical catalog record (`reference-lock-v1`). */
export const catalogDigest = (
  raw: Readonly<Record<string, unknown>>,
): Effect.Effect<string, CatalogError, Crypto.Crypto> =>
  Effect.gen(function* () {
    const crypto = yield* Crypto.Crypto;
    const canonical = yield* Effect.try({
      try: () => canonicalizeForDigest(raw),
      catch: (cause) =>
        new CatalogError({ message: "cannot canonicalize catalog record for digest", cause }),
    });
    const digest = yield* crypto
      .digest("SHA-256", new TextEncoder().encode(canonical))
      .pipe(
        Effect.mapError(
          (cause) => new CatalogError({ message: "cannot compute catalog digest", cause }),
        ),
      );
    return toHex(digest);
  });

export interface Catalog {
  readonly sources: ReadonlyMap<string, CatalogSource>;
}

const SourceRecordSchema = Schema.Struct({
  id: Schema.NonEmptyString,
  kind: Schema.NonEmptyString,
  origin: Schema.NonEmptyString,
  local_hint: Schema.optionalKey(Schema.NonEmptyString),
  origin_aliases: Schema.optionalKey(Schema.NonEmptyArray(Schema.String)),
  track: Schema.optionalKey(Schema.NonEmptyString),
  license_paths: Schema.optionalKey(Schema.NonEmptyArray(Schema.String)),
  classes: Schema.optionalKey(Schema.NonEmptyArray(Schema.String)),
  questions: Schema.optionalKey(Schema.NonEmptyArray(Schema.String)),
});

const validateRecord = (
  record: Record<string, unknown>,
): Effect.Effect<CatalogSource, CatalogError> =>
  Effect.gen(function* () {
    const fields = yield* Schema.decodeUnknownEffect(SourceRecordSchema)(record).pipe(
      Effect.mapError(
        (cause) => new CatalogError({ message: `invalid source record: ${cause.message}`, cause }),
      ),
    );

    if (!isValidSourceId(fields.id)) {
      return yield* new CatalogError({
        message:
          `source id ${JSON.stringify(fields.id)} is not a path-safe dotted identifier ` +
          "(expected lowercase alphanumeric/hyphen segments joined by single dots)",
      });
    }
    if (!isGitSafeValue(fields.origin)) {
      return yield* new CatalogError({
        message: `source ${JSON.stringify(fields.id)}: 'origin' is not safe (option-like or has control characters)`,
      });
    }
    const originAliases = fields.origin_aliases ?? [];
    for (const alias of originAliases) {
      if (!isGitSafeValue(alias)) {
        return yield* new CatalogError({
          message: `source ${JSON.stringify(fields.id)}: origin_aliases entry ${JSON.stringify(alias)} is not safe`,
        });
      }
    }
    const track = fields.track ?? null;
    if (track !== null && !isGitSafeValue(track)) {
      return yield* new CatalogError({
        message: `source ${JSON.stringify(fields.id)}: 'track' is not safe (option-like or has control characters)`,
      });
    }
    const licensePathsRaw = fields.license_paths ?? [];
    for (const path of licensePathsRaw) {
      if (!isValidLicensePath(path)) {
        return yield* new CatalogError({
          message: `source ${JSON.stringify(fields.id)}: license path ${JSON.stringify(path)} is not a normalized relative path`,
        });
      }
    }
    if (new Set(licensePathsRaw).size !== licensePathsRaw.length) {
      return yield* new CatalogError({
        message: `source ${JSON.stringify(fields.id)}: license_paths contains duplicates`,
      });
    }
    if ((track === null) !== (licensePathsRaw.length === 0)) {
      return yield* new CatalogError({
        message:
          `source ${JSON.stringify(fields.id)}: 'track' and 'license_paths' must be declared ` +
          "together (a source is either fully lockable or fully unlocked)",
      });
    }

    const source: CatalogSource = {
      id: fields.id,
      kind: fields.kind,
      origin: fields.origin,
      localHint: fields.local_hint ?? null,
      originAliases,
      track,
      licensePaths: licensePathsRaw,
      classes: fields.classes ?? [],
      questions: fields.questions ?? [],
      raw: record,
    };
    return source;
  });

const DocumentSchema = Schema.Struct({
  schema: Schema.Unknown,
  source: Schema.optionalKey(Schema.Array(Schema.Unknown)),
});

export const parseCatalogText = (text: string): Effect.Effect<Catalog, CatalogError, TomlParser> =>
  Effect.gen(function* () {
    const parser = yield* TomlParser;
    const parsed = yield* parser
      .parse(text)
      .pipe(
        Effect.mapError(
          (cause) =>
            new CatalogError({ message: `catalog is not valid TOML: ${cause.message}`, cause }),
        ),
      );
    const document = yield* Schema.decodeUnknownEffect(DocumentSchema)(parsed).pipe(
      Effect.mapError(
        (cause) =>
          new CatalogError({
            message: `catalog document has an invalid shape: ${cause.message}`,
            cause,
          }),
      ),
    );
    if (document.schema !== 1) {
      return yield* new CatalogError({
        message: `catalog schema ${JSON.stringify(document.schema)} is not the supported value (1)`,
      });
    }

    const sources = new Map<string, CatalogSource>();
    for (const record of document.source ?? []) {
      if (typeof record !== "object" || record === null || Array.isArray(record)) {
        return yield* new CatalogError({ message: "each [[source]] entry must be a table" });
      }
      const source = yield* validateRecord(record as Record<string, unknown>);
      if (sources.has(source.id)) {
        return yield* new CatalogError({
          message: `duplicate source id ${JSON.stringify(source.id)}`,
        });
      }
      sources.set(source.id, source);
    }
    return { sources };
  });

export const loadCatalog = (
  path: string,
): Effect.Effect<Catalog, CatalogError, TomlParser | FileSystem.FileSystem> =>
  Effect.gen(function* () {
    const fs = yield* FileSystem.FileSystem;
    const text = yield* fs
      .readFileString(path)
      .pipe(
        Effect.mapError(
          (cause) => new CatalogError({ message: `cannot read catalog at ${path}`, cause }),
        ),
      );
    return yield* parseCatalogText(text);
  });
